flight/Modules/Actuator/actuator.c

```diff
@@ -218,282 +231,416 @@ static float get_curve2_source(ActuatorDesiredData *desired, SystemSettingsAirfr
+				num_motors++;
+				break;
+			case MIXERSETTINGS_MIXER1TYPE_CAMERAPITCH:
+				if (CameraDesiredHandle()) {
+					CameraDesiredPitchGet(
+							&motor_vect[ct]);
+				} else {
+					motor_vect[ct] = -1;
+				}
+				break;
+			case MIXERSETTINGS_MIXER1TYPE_CAMERAROLL:
+				if (CameraDesiredHandle()) {
+					CameraDesiredRollGet(
+							&motor_vect[ct]);
+				} else {
+					motor_vect[ct] = -1;
+				}
+				break;
+			case MIXERSETTINGS_MIXER1TYPE_CAMERAYAW:
+				if (CameraDesiredHandle()) {
+					CameraDesiredRollGet(
+							&motor_vect[ct]);
+				} else {
+					motor_vect[ct] = -1;
+				}
+				break;
+			default:
+				set_failsafe();
+				PIOS_Assert(0);
 		}
+	}
 
-		if (rc != true) {
-			/* Update of ActuatorDesired timed out,
-			 * or first iteration.  Go to failsafe */
-			set_failsafe();
-		}
+	float gain = 1.0f;
+	float offset = 0.0f;
 
-		PIOS_WDG_UpdateFlag(PIOS_WDG_ACTUATOR);
+	/* This is a little dubious.  Scale down command ranges to
+	 * fit.  It may cause some cross-axis coupling, though
+	 * generally less than if we were to actually let it clip.
+	 */
+	if ((max_chan - min_chan) > 1.0f) {
+		gain = 1.0f / (max_chan - min_chan);
 
-		UAVObjEvent ev;
+		max_chan *= gain;
+		min_chan *= gain;
+	}
 
-		// Wait until the ActuatorDesired object is updated
-		rc = PIOS_Queue_Receive(queue, &ev, FAILSAFE_TIMEOUT_MS);
+	/* Sacrifice throttle because of clipping */
+	if (max_chan > 1.0f) {
+		offset = 1.0f - max_chan;
+	} else if (min_chan < 0.0f) {
+		/* Low-side clip management-- how much power are we
+		 * willing to add??? */
+
+		neg_clip /= num_motors;
+
+		/* neg_clip is now the amount of throttle "already added." by
+		 * clipping...
+		 *
+		 * Find the "highest possible value" of offset.
+		 * if neg_clip is -15%, and maxpoweradd is 10%, we need to add
+		 * -5% to all motors.
+		 * if neg_clip is 5%, and maxpoweradd is 10%, we can add up to
+		 * 5% to all motors to further fix clipping.
+		 */
+		offset = neg_clip + actuatorSettings.LowPowerStabilizationMaxPowerAdd;
 
-		/* If we timed out, go to top of loop, which sets failsafe
-		 * and waits again. */
-		if (rc != true) {
-			continue;
+		/* Add the lesser of--
+		 * A) the amount the lowest channel is out of range.
+		 * B) the above calculated offset.
+		 */
+		offset = MIN(-min_chan, offset);
+	}
+
+	for (int ct = 0; ct < MAX_MIX_ACTUATORS; ct++) {
+		// Motors have additional protection for when to be on
+		if (types_mixer[ct] == MIXERSETTINGS_MIXER1TYPE_MOTOR) {
+			if (!armed) {
+				motor_vect[ct] = -1;  //force min throttle
+			} else if (!stabilize_now) {
+				if (!spin_while_armed) {
+					motor_vect[ct] = -1;
+				} else {
+					motor_vect[ct] = 0;
+				}
+			} else {
+				motor_vect[ct] = motor_vect[ct] * gain + offset;
+
+				if (motor_vect[ct] > 0) {
+					// Apply curve fitting, mapping the input to the propeller output.
+					motor_vect[ct] = powapprox(motor_vect[ct], actuatorSettings.MotorInputOutputCurveFit);
```

Comment:
This is the one thing that is complicated for actuator feedback, etc.   This should ultimately get deferred to happen "later".